File: sdks/dotnet/src/Rrflow.Rrd.Client/RetryPolicy.cs
using System;

namespace Rrflow.Rrd;

internal sealed class RetryPolicy
{
    private readonly TimeSpan _requestTimeout;
    private readonly int _maxAttempts;

    internal RetryPolicy(ClientOptions options)
    {
        _requestTimeout = options.RequestTimeout;
        _maxAttempts = options.MaxAttempts;
    }

    internal int AttemptsFor(Endpoint endpoint, RequestOptions options)
    {
        bool retrySafe = endpoint.Method == "GET"
            || !endpoint.Mutation
            || options.IdempotencyKey is not null;
        return retrySafe ? _maxAttempts : 1;
    }

    internal TimeSpan RemainingTimeout(DateTimeOffset? deadline)
    {
        if (deadline is null)
        {
            return _requestTimeout;
        }
        TimeSpan remaining = deadline.Value - DateTimeOffset.UtcNow;
        if (remaining <= TimeSpan.Zero)
        {
            throw new RrdClientException("RRD request deadline has expired");
        }
        return remaining < _requestTimeout ? remaining : _requestTimeout;
    }
}

File: sdks/java/src/main/java/io/rrflow/rrd/RetryPolicy.java
package io.rrflow.rrd;

import java.time.Duration;
import java.time.Instant;

/** Current broad replay and deadline behavior; H-04 owns semantic certainty. */
final class RetryPolicy {
    private final Duration requestTimeout;
    private final int maxAttempts;

    RetryPolicy(Duration requestTimeout, int maxAttempts) {
        this.requestTimeout = requestTimeout;
        this.maxAttempts = maxAttempts;
    }

    int attempts(OperationId operation, RequestOptions options) {
        boolean retrySafe = operation.method().equals("GET")
                || !operation.mutation()
                || options.idempotencyKey() != null;
        return retrySafe ? maxAttempts : 1;
    }

    Duration remainingTimeout(Instant deadline) {
        if (deadline == null) {
            return requestTimeout;
        }
        Duration remaining = Duration.between(Instant.now(), deadline);
        if (remaining.isZero() || remaining.isNegative()) {
            throw new RrdClientException("RRD request deadline has expired");
        }
        return remaining.compareTo(requestTimeout) < 0 ? remaining : requestTimeout;
    }
}
